### PR TITLE
plan: consider task tp in required prop.

### PR DESCRIPTION
**Pull request by hanfei1991**
For SQL: `select * from t where a < 10`, the plan is `DataSource -> Projection`. If the selectivity of a is 0.8, we are sure to choose table scan. But DataSource will not finish the index look up task when comparing the table scan and index look up. So Index look up has less cost than it should take. We should add the task type in required property to get the right plan.
@shenli @zimulala @winoros @lamxTyler PTAL

### CODE REVIEW

**Comment by shenli** [comment]
> But DataSource will not finish the index look up task when comparing the table scan and index look up. So Index look up has less cost than it should take. 
This description confuses me.

File: plan/new_physical_plan_builder.go [hunk 1]
```diff
@@ -27,6 +27,10 @@ import (
 	"github.com/pingcap/tidb/util/types"
 )
 
+// taskTypes records all possible kinds of task that a plan can return. For Agg, TopN and Limit, we will try to get
+// these tasks one by one.
+var taskTypes = []taskType{rootTask, copSingleReadTask, copDoubleReadTask}
```
Comment (shenli) [inline plan/new_physical_plan_builder.go]: There are only three kinds of tasks?

File: plan/new_physical_plan_builder.go [hunk 2]
```diff
@@ -259,21 +273,26 @@ func (p *Sort) convert2NewPhysicalPlan(prop *requiredProp) (taskProfile, error)
 	if task != nil {
 		return task, nil
 	}
+	if prop.taskTp != rootTask {
+		// TODO: This is a trick here, because a operator that can be pushed to Coprocessor can never be pushed across sort.
```
Comment (shenli) [inline plan/new_physical_plan_builder.go]: a -> an

File: plan/new_physical_plan_builder.go (hunk 1 above)
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: currently, it is

**Comment by hanfei1991** [comment]
@shenli PTAL

File: plan/new_physical_plan_builder.go [hunk 3]
```diff
@@ -77,6 +81,11 @@ func (p *Projection) convert2NewPhysicalPlan(prop *requiredProp) (taskProfile, e
 	if task != nil {
 		return task, nil
 	}
+	if prop.taskTp != rootTask {
+		// Projection cannot be pushed down currently, it can only return rootTask.
```
Comment (zimulala) [inline plan/new_physical_plan_builder.go]: Complete this comments or update this code.

File: plan/new_physical_plan_builder.go [hunk 4]
```diff
@@ -109,6 +118,11 @@ func (p *LogicalJoin) convert2NewPhysicalPlan(prop *requiredProp) (taskProfile,
 	if task != nil {
 		return task, nil
 	}
+	if prop.taskTp != rootTask {
+		// Join cannot be pushed down currently, it can only return rootTask.
```
Comment (zimulala) [inline plan/new_physical_plan_builder.go]: Ditto.

File: plan/plan.go [hunk 5]
```diff
@@ -75,16 +75,47 @@ func (c *columnProp) equal(nc *columnProp, ctx context.Context) bool {
 	return c.col.Equal(nc.col, ctx) && c.desc == nc.desc
 }
 
+// taskType is the type of execution task.
+type taskType int
+
+const (
+	rootTask          taskType = iota
```
Comment (zimulala) [inline plan/plan.go]: s/rootTask/rootTaskType

File: plan/new_physical_plan_builder.go [hunk 6]
```diff
@@ -723,11 +815,16 @@ func (p *LogicalApply) convert2NewPhysicalPlan(prop *requiredProp) (taskProfile,
 	if task != nil {
 		return task, nil
 	}
+	if prop.taskTp != rootTask {
+		// Apply can only return rootTask.
+		task = &copTaskProfile{cst: math.MaxFloat64}
+		return task, p.storeTaskProfile(prop, task)
+	}
 	// TODO: refine this code.
```
Comment (zimulala) [inline plan/new_physical_plan_builder.go]: s/refine/Refine

File: plan/new_physical_plan_builder.go (hunk 1 above)
Comment (zimulala) [inline plan/new_physical_plan_builder.go]: s/taskTypes/allTaskTypes

**Comment by XuHuaiyu** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

**Comment by zimulala** [comment]
LGTM

**Pull request merged**